src/pint/models/stand_alone_psr_binaries/BT_piecewise.py

```diff
@@ -0,0 +1,441 @@
+        else:
+            T0 = self.T0
+            # print("hello from 4")
+        if not hasattr(barycentricTOA, "unit") or barycentricTOA.unit == None:
+            barycentricTOA = barycentricTOA * u.day
+        # print(f"Unique T0s being used in tt0 calculation: {np.unique(T0)}\n")
+        tt0 = (barycentricTOA - T0).to("second")
+        # print(f"piecewise tt0 being called, Calc'd with: {T0} ")
+        return tt0
+
+    def d_delayL1_d_par(self, par):
+        if par not in self.binary_params:
+            errorMesg = par + " is not in binary parameter list."
+            raise ValueError(errorMesg)
+        par_obj = getattr(self, par)
+        index, par_temp = self.in_piece(par)
+        # print(index)
+        if (
+            par_temp is None
+        ):  # to circumvent the need for list of d_pars = [T0X_0,...,T0X_i] use par_temp
+            if hasattr(self, "d_delayL1_d_" + par):
+                func = getattr(self, "d_delayL1_d_" + par)
+                return func() * index
+            else:
+                if par in self.orbits_cls.orbit_params:
+                    return self.d_delayL1_d_E() * self.d_E_d_par(par)
+                else:
+                    return np.zeros(len(self.t)) * u.second / par_obj.unit
+        else:
+            if hasattr(self, "d_delayL1_d_" + par_temp):
+                func = getattr(self, "d_delayL1_d_" + par_temp)
+                return func() * index
+            else:
+                if par in self.orbits_cls.orbit_params:
+                    return self.d_delayL1_d_E() * self.d_E_d_par()
+                else:
+                    return np.zeros(len(self.t)) * u.second / par_obj.unit
+
+    def d_delayL2_d_par(self, par):
+        if par not in self.binary_params:
+            errorMesg = par + " is not in binary parameter list."
+            raise ValueError(errorMesg)
+        # print(par)
+        par_obj = getattr(self, par)
+        index, par_temp = self.in_piece(par)
+        # print(index)
+        if (
+            par_temp is None
+        ):  # to circumvent the need for list of d_pars = [T0X_0,...,T0X_i] use par_temp
+            if hasattr(self, "d_delayL2_d_" + par):
+                func = getattr(self, "d_delayL2_d_" + par)
+                return func() * index
+            else:
+                if par in self.orbits_cls.orbit_params:
+                    return self.d_delayL2_d_E() * self.d_E_d_par(par)
+                else:
+                    return np.zeros(len(self.t)) * u.second / par_obj.unit
+        else:
+            if hasattr(self, "d_delayL2_d_" + par_temp):
+                func = getattr(self, "d_delayL2_d_" + par_temp)
+                return func() * index
+            else:
+                if par in self.orbits_cls.orbit_params:
+                    return self.d_delayL2_d_E() * self.d_E_d_par()
+                else:
+                    return np.zeros(len(self.t)) * u.second / par_obj.unit
+
+    def in_piece(self, par):
```

Comment:
I think this function needs a docstring - what does it do?

Reply:
Added doc string to in_piece.
It finds which TOAs reference a given piecewise binary parameter group par.
It looks up these values internally from the property group_array_index, which is generated when `self._t` is not `None`
Returns a boolean list (length = number of TOAs), True for the index where TOAs reference a given piecewise parameter, False otherwise. 

Just occurred to me that this may be an issue if the TOAs are reordered between calculations. Unless the TOAs always reach here in a controlled order